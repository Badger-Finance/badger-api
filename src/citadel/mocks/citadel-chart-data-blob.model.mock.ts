import { ChartDataBlob } from '../../aws/models/chart-data-blob.model';
import { HistoricTreasurySummarySnapshot } from '../../aws/models/historic-treasury-summary-snapshot.model';

export const citadelChartDataBlobModel: ChartDataBlob<HistoricTreasurySummarySnapshot> = Object.assign(
  new ChartDataBlob<HistoricTreasurySummarySnapshot>(),
  {
    id: 'treasury_0x38724146C8dc1Aa49c3395091cf86B789c37F52c_1y',
    timeframe: '1y',
    data: [
      {
        positions: [
          {
            symbol: 'BADGER',
            apr: 0,
            protocol: 'none',
            address: '0x3472A5A71965499acd81997a54BBA8D852C6E53d',
            balance: 4,
            decimals: 18,
            name: 'Badger',
            value: 23.4,
          },
          {
            symbol: 'WBTC',
            apr: 0,
            protocol: 'none',
            address: '0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599',
            balance: 0.002,
            decimals: 8,
            name: 'Wrapped BTC',
            value: 60.194,
          },
          {
            symbol: 'WETH',
            apr: 0,
            protocol: 'none',
            address: '0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2',
            balance: 0.025,
            decimals: 18,
            name: 'Wrapped Ether',
            value: 50.41375,
          },
          {
            symbol: 'USDC',
            apr: 0,
            protocol: 'none',
            address: '0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48',
            balance: 50,
            decimals: 6,
            name: 'USD Coin',
            value: 50.14999999999999,
          },
          {
            symbol: 'FRAX',
            apr: 0,
            protocol: 'none',
            address: '0x853d955aCEf822Db058eb8505911ED77F175b99e',
            balance: 50,
            decimals: 18,
            name: 'Frax',
            value: 50.05,
          },
          {
            symbol: 'UST',
            apr: 0,
            protocol: 'none',
            address: '0xa47c8bf37f92aBed4A126BDA807A7b7498661acD',
            balance: 62.97449812052163,
            decimals: 18,
            name: 'Wrapped UST Token',
            value: 4.013112867228362,
          },
          {
            symbol: 'CVX',
            apr: 0,
            protocol: 'none',
            address: '0x4e3FBD56CD56c3e72c1403e103b45Db9da5B9D2B',
            balance: 1.00003e-13,
            decimals: 18,
            name: 'Convex Token',
            value: 1.02803084e-12,
          },
          {
            symbol: 'bcvxCRV',
            apr: 57.56724452641197,
            protocol: 'Badger',
            address: '0x2B5455aac8d64C14786c3a29858E43b5945819C0',
            balance: 1,
            decimals: 18,
            name: 'Badger Sett Convex CRV',
            value: 1.7347946415600242,
          },
          {
            symbol: 'bveCVX',
            apr: 31.48681878583217,
            protocol: 'Badger',
            address: '0xfd05D3C7fe2924020620A8bE4961bBaA747e6305',
            balance: 2,
            decimals: 18,
            name: 'Badger Vested Escrow Convex Token',
            value: 20.5,
          },
          {
            symbol: 'bibbtc/sbtcCRV-f',
            apr: 139.74480236006133,
            protocol: 'Badger',
            address: '0xaE96fF08771a109dc6650a1BdCa62F2d558E40af',
            balance: 0.0005,
            decimals: 18,
            name: 'Badger Sett Curve.fi Factory BTC Metapool: ibBTC',
            value: 14.889253336268698,
          },
        ],
        address: '0x38724146C8dc1Aa49c3395091cf86B789c37F52c',
        id: '0x38724146C8dc1Aa49c3395091cf86B789c37F52c',
        value: 275.3449108450581,
        yield: 10.263646742319406,
        timestamp: 1653251755438,
      },
      {
        positions: [
          {
            symbol: 'BADGER',
            apr: 0,
            protocol: 'none',
            address: '0x3472A5A71965499acd81997a54BBA8D852C6E53d',
            balance: 4,
            decimals: 18,
            name: 'Badger',
            value: 23.2,
          },
          {
            symbol: 'WBTC',
            apr: 0,
            protocol: 'none',
            address: '0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599',
            balance: 0.002,
            decimals: 8,
            name: 'Wrapped BTC',
            value: 58.998000000000005,
          },
          {
            symbol: 'WETH',
            apr: 0,
            protocol: 'none',
            address: '0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2',
            balance: 0.025,
            decimals: 18,
            name: 'Wrapped Ether',
            value: 49.556000000000004,
          },
          {
            symbol: 'USDC',
            apr: 0,
            protocol: 'none',
            address: '0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48',
            balance: 50,
            decimals: 6,
            name: 'USD Coin',
            value: 50.1,
          },
          {
            symbol: 'FRAX',
            apr: 0,
            protocol: 'none',
            address: '0x853d955aCEf822Db058eb8505911ED77F175b99e',
            balance: 50,
            decimals: 18,
            name: 'Frax',
            value: 49.9753,
          },
          {
            symbol: 'UST',
            apr: 0,
            protocol: 'none',
            address: '0xa47c8bf37f92aBed4A126BDA807A7b7498661acD',
            balance: 62.97449812052163,
            decimals: 18,
            name: 'Wrapped UST Token',
            value: 3.143889724060388,
          },
          {
            symbol: 'CVX',
            apr: 0,
            protocol: 'none',
            address: '0x4e3FBD56CD56c3e72c1403e103b45Db9da5B9D2B',
            balance: 1.00003e-13,
            decimals: 18,
            name: 'Convex Token',
            value: 9.560286800000002e-13,
          },
          {
            symbol: 'bcvxCRV',
            apr: 57.324597179856944,
            protocol: 'Badger',
            address: '0x2B5455aac8d64C14786c3a29858E43b5945819C0',
            balance: 1,
            decimals: 18,
            name: 'Badger Sett Convex CRV',
            value: 1.6914247755210234,
          },
          {
            symbol: 'bveCVX',
            apr: 32.27236423063117,
            protocol: 'Badger',
            address: '0xfd05D3C7fe2924020620A8bE4961bBaA747e6305',
            balance: 2,
            decimals: 18,
            name: 'Badger Vested Escrow Convex Token',
            value: 19.12,
          },
          {
            symbol: 'bibbtc/sbtcCRV-f',
            apr: 139.96848608236724,
            protocol: 'Badger',
            address: '0xaE96fF08771a109dc6650a1BdCa62F2d558E40af',
            balance: 0.0005,
            decimals: 18,
            name: 'Badger Sett Curve.fi Factory BTC Metapool: ibBTC',
            value: 14.784913808979702,
          },
        ],
        address: '0x38724146C8dc1Aa49c3395091cf86B789c37F52c',
        id: '0x38724146C8dc1Aa49c3395091cf86B789c37F52c',
        value: 270.5695283085621,
        yield: 10.28729978615088,
        timestamp: 1653165354948,
      },
    ],
  },
);
