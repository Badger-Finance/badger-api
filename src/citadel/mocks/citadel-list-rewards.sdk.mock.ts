import { ListRewardsEvent } from '@badger-dao/sdk/lib/citadel/interfaces/list-rewards-event.interface';
import { RewardFilter } from '@badger-dao/sdk/lib/citadel/enums/reward-filter.enum';
import { BigNumber } from 'ethers';

export const citadelListRewardsSdkMock: ListRewardsEvent[] = [
  {
    account: '0xA967Ba66Fb284EC18bbe59f65bcf42dD11BA8128',
    block: 50,
    dataTypeHash: '0x60b54c1986b43275418006baba637cb29964609f76738f9d0c47dbd9c076bf11',
    reward: BigNumber.from(64),
    timestamp: 1651088758542,
    token: '0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599',
    type: RewardFilter.ADDED,
  },
  {
    account: '0xA967Ba66Fb284EC18bbe59f65bcf42dD11BA8128',
    block: 50,
    dataTypeHash: '0x60b54c1986b43275418006baba637cb29964609f76738f9d0c47dbd9c076bf11',
    reward: BigNumber.from(8),
    timestamp: 1651088758542,
    token: '0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599',
    type: RewardFilter.ADDED,
  },
  {
    account: '0xA967Ba66Fb284EC18bbe59f65bcf42dD11BA8128',
    block: 51,
    dataTypeHash: '0x60b54c1986b43275418006baba637cb29964609f76738f9d0c47dbd9c076bf11',
    reward: BigNumber.from(33),
    timestamp: 1651088758542,
    token: '0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599',
    type: RewardFilter.ADDED,
  },
  {
    account: '0xA967Ba66Fb284EC18bbe59f65bcf42dD11BA8128',
    block: 52,
    dataTypeHash: '0x60b54c1986b43275418006baba637cb29964609f76738f9d0c47dbd9c076bf11',
    reward: BigNumber.from(123),
    timestamp: 1651088758542,
    token: '0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599',
    type: RewardFilter.ADDED,
  },
  {
    account: '0xA967Ba66Fb284EC18bbe59f65bcf42dD11BA8128',
    block: 52,
    dataTypeHash: '0x60b54c1986b43275418006baba637cb29964609f76738f9d0c47dbd9c076bf11',
    reward: BigNumber.from(14),
    timestamp: 1651088758542,
    token: '0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599',
    type: RewardFilter.ADDED,
  },
];
