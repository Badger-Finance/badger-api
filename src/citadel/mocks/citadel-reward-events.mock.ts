import { RewardFilter } from '@badger-dao/sdk/lib/citadel/enums/reward-filter.enum';
import { TOKENS } from '../../config/tokens.config';
import { CitadelRewardEvent } from '../interfaces/citadel-reward-event.interface';

const TEST_TOKEN = TOKENS.CTDL;
const TEST_USER = '0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28';

export const citadelRewardEventsMock: CitadelRewardEvent[] = [
  {
    block: 10,
    user: TEST_USER,
    token: TEST_TOKEN,
    amount: 10.234,
    type: RewardFilter.ADDED,
  },
  {
    block: 11,
    user: TEST_USER,
    token: TEST_TOKEN,
    amount: 3,
    type: RewardFilter.ADDED,
  },
  {
    block: 12,
    user: TEST_USER,
    token: TEST_TOKEN,
    amount: 0,
    type: RewardFilter.ADDED,
  },
];
