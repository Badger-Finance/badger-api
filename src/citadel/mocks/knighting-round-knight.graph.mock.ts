import { KnightsRoundsQuery, VotesQuery } from '../../graphql/generated/citadel.knights.round';

export const knightingRoundKnightGraphMock: KnightsRoundsQuery = {
  knights: [
    {
      id: '0',
      voters: [
        {
          id: '00xef42d748e09a2d9ef89238c053ce0b6f00236210',
        },
      ],
      votes: [
        {
          id: '0x6961d11d4fdafabebb616eb9bcd163a0f1528774112375b612d0ffc63a507b4b',
        },
        {
          id: '0xdc1a188d0e557dffbf9ad40b1371262273b1a754c48251d6cadcdd7e0255dc05',
        },
      ],
      voteCount: '2',
      voteAmount: '3',
    },
  ],
};

export const knightingRoundVoteGraphMock: VotesQuery = {
  votes: [
    {
      id: '0x3a4fd25adb4ff9940ff898915f15cc589a00021d02c867320e613dcca2a7e078',
      knight: {
        id: '2',
      },
      voter: {
        id: '0xa967ba66fb284ec18bbe59f65bcf42dd11ba8128',
      },
      amount: '276190476190000000',
    },
    {
      id: '0x8fcf54ccfd681f3ceef70b8b96e14db1c95cea9b864765d162847f4137dbabb4',
      knight: {
        id: '5',
      },
      voter: {
        id: '0xc2e345f74b18187e5489822f9601c028ed1915a2',
      },
      amount: '95238095220000000',
    },
    {
      id: '0x92a4ab357457f40c670785bac023176d4c06a9bc9ab3180939cae61c1429df4a',
      knight: {
        id: '3',
      },
      voter: {
        id: '0xc2e345f74b18187e5489822f9601c028ed1915a2',
      },
      amount: '276190476190000000',
    },
    {
      id: '0xb9d1992f8d2cce278ba7e6089af2dbb1e82a8a6a6fdc803417ce5f01719122ff',
      knight: {
        id: '5',
      },
      voter: {
        id: '0x17879d10c22f17faa0b2a29a36e968712f8127c9',
      },
      amount: '276190476190000000',
    },
    {
      id: '0xf5a479704f66978e5c101aa67beaf8adff8ca6c2603948b1c064d76764922a3e',
      knight: {
        id: '4',
      },
      voter: {
        id: '0xa967ba66fb284ec18bbe59f65bcf42dd11ba8128',
      },
      amount: '999999999810000000',
    },
    {
      id: '0xfd553615365db370df07f2231afcd938e66b27c8f419595277a5d8982e58ae3b',
      knight: {
        id: '7',
      },
      voter: {
        id: '0xc2e345f74b18187e5489822f9601c028ed1915a2',
      },
      amount: '190476190440000000',
    },
  ],
};
