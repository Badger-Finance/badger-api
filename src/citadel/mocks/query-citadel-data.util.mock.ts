import { CitadelData } from '../destructors/citadel-data.destructor';

const citadelDataBlob = new Map();

citadelDataBlob.set('stakedPercent', 57.21943985085593);
citadelDataBlob.set('marketCap', 309486236.0874132);
citadelDataBlob.set('marketCapToTreasuryRatio', 1126133.0423130456);
citadelDataBlob.set('tokensPaid', {});
citadelDataBlob.set('valuePaid', 0);
citadelDataBlob.set('staked', 123433.75281462506);
citadelDataBlob.set('supply', 215719.96009810406);
citadelDataBlob.set('stakingBps', 3000);
citadelDataBlob.set('fundingBps', 4000);
citadelDataBlob.set('lockingBps', 3000);
citadelDataBlob.set('stakingApr', 2498.4764247727794);
citadelDataBlob.set('lockingApr', {
  '0': 0,
  '1': 0,
  '2': 0,
  '3': 1672009.2017791115,
  overall: 1672009.2017791115,
});

export const queryCitadelDataUtilMock: CitadelData = new CitadelData(citadelDataBlob);
