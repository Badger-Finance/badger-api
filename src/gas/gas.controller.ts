import { GasPrices, Network } from '@badger-dao/sdk';
import { Controller, Get, QueryParams, UseCache } from '@tsed/common';
import { ContentType, Description, Returns, Summary } from '@tsed/schema';

import { Chain } from '../chains/config/chain.config';

@Controller('/gas')
export class GasController {
  @Get('')
  @UseCache()
  @ContentType('json')
  @Summary('Get the current gas price')
  @Description('Returns the current gas price on the requested chain')
  @Returns(200)
  @Returns(404).Description('Chain gas prices not available')
  async getGasPrices(@QueryParams('chain') chain?: Network): Promise<GasPrices> {
    const targetChain = Chain.getChain(chain);
    return targetChain.getGasPrices();
  }
}
