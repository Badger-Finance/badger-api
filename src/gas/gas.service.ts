import { Service } from '@tsed/common';

import { Chain } from '../chains/config/chain.config';
import { getGasCache } from './gas.utils';
import { GasPrices } from './interfaces/gas-prices.interface';

@Service()
export class GasService {
  /**
   * Attempt to retrieve the prices for a chain from the cache
   * if no value is in the cache, refresh prices for all chains
   * @param chain Gas price chain
   * @returns object of gas speeds and prices
   */
  async getGasPrices(chain: Chain): Promise<GasPrices> {
    const gasCache = await getGasCache();
    return gasCache[chain.network];
  }
}
