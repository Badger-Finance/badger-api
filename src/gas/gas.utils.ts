import NodeCache from 'node-cache';

import { SUPPORTED_CHAINS } from '../chains/chain';
import { ChainGasPrices } from './interfaces/gas-prices.interface';

const gasCache = new NodeCache({ stdTTL: 15, checkperiod: 17 });

export async function getGasCache(): Promise<ChainGasPrices> {
  const cachedGasPrices = gasCache.get<ChainGasPrices>('gasPrices');
  if (cachedGasPrices) {
    return cachedGasPrices;
  }
  const chainGasPrices = await Promise.all(SUPPORTED_CHAINS.map(async (c) => [c.network, await c.getGasPrices()]));
  const gasPrices = Object.fromEntries(chainGasPrices);
  gasCache.set('gasPrices', gasPrices);
  return gasPrices;
}
