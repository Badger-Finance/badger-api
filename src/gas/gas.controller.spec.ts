import { PlatformServerless } from "@tsed/platform-serverless";
import { PlatformServerlessTest } from "@tsed/platform-serverless-testing";

import { setupMockChain } from "../test/mocks.utils";
import { GasController } from "./gas.controller";

describe("GasController", () => {
  beforeEach(
    PlatformServerlessTest.bootstrap(PlatformServerless, {
      lambda: [GasController]
    })
  );
  afterEach(() => PlatformServerlessTest.reset());

  beforeEach(setupMockChain);

  describe("GET /gas", () => {
    it("returns gas price options in eip-1559 format", async () => {
      const { body, statusCode } = await PlatformServerlessTest.request.get("/gas");
      expect(statusCode).toEqual(200);
      expect(JSON.parse(body)).toMatchSnapshot();
    });
  });
});
