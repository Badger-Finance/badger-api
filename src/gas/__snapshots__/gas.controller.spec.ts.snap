// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GasController GET /v2/gas returns gas price options in eip-1559 format 1`] = `
Object {
  "fast": Object {
    "maxFeePerGas": 221.96,
    "maxPriorityFeePerGas": 1.94,
  },
  "rapid": Object {
    "maxFeePerGas": 223.06,
    "maxPriorityFeePerGas": 3.04,
  },
  "slow": Object {
    "maxFeePerGas": 221.81,
    "maxPriorityFeePerGas": 1.79,
  },
  "standard": Object {
    "maxFeePerGas": 221.91,
    "maxPriorityFeePerGas": 1.89,
  },
}
`;
